compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/UpdateCyclicIntersection.java

```diff
@@ -0,0 +1,467 @@
+import org.wso2.ballerinalang.compiler.semantics.model.types.BRecordType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BStructureType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTableType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTupleMember;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTupleType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BXMLType;
+import org.wso2.ballerinalang.compiler.tree.BLangNode;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.ImmutableTypeCloner;
+import org.wso2.ballerinalang.compiler.util.Name;
+import org.wso2.ballerinalang.compiler.util.Names;
+import org.wso2.ballerinalang.compiler.util.TypeTags;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+import static org.ballerinalang.model.symbols.SymbolOrigin.SOURCE;
+
+/**
+ * @since 2201.7.0
+ */
+public class UpdateCyclicIntersection implements TypeVisitor {
+
+    private static final CompilerContext.Key<UpdateCyclicIntersection> UPDATE_IMMUTABLE_TYPE_KEY =
+                new CompilerContext.Key<>();
+    private final SymbolTable symTable;
+    private final Names names;
+    private final Types types;
+    private final BLangAnonymousModelHelper anonymousModelHelper;
+    private Location loc;
+    private PackageID pkgID;
+    private BLangNode typeNode;
+    private SymbolEnv env;
+    private HashSet<BType> visitedImmutableTypes = new HashSet<>();
+
+    public UpdateCyclicIntersection(CompilerContext context) {
+        context.put(UPDATE_IMMUTABLE_TYPE_KEY, this);
+        this.symTable = SymbolTable.getInstance(context);
+        this.names = Names.getInstance(context);
+        this.types = Types.getInstance(context);
+        this.anonymousModelHelper = BLangAnonymousModelHelper.getInstance(context);
+    }
+
+    public static UpdateCyclicIntersection getInstance(CompilerContext context) {
+        UpdateCyclicIntersection updateImmutableType = context.get(UPDATE_IMMUTABLE_TYPE_KEY);
+        if (updateImmutableType == null) {
+            updateImmutableType = new UpdateCyclicIntersection(context);
+        }
+        return updateImmutableType;
+    }
+
+    public void updateType(BType immutableType, Location loc, PackageID pkgID, BLangNode typeNode,
+                           SymbolEnv env) {
+        if (!visitedImmutableTypes.add(immutableType)) {
+            return;
+        }
```

Comment:
I think we need to calculate the correct effective type here using the constituent types and create overloaded methods to accept the old type and the correct type and populate the each type with with the correct one. Seems this will require overloaded methods, not a visitor.

Reply:
@hasithaa WDYT?